[Docs] Refine client-side admin policy doc and example (#6089)
skypilot-org/skypilot: docs/source/cloud-setup/policy.rst

To deploy a admin policy, here are the steps:

1. :ref:`Implement the policy in a Python package <implement-admin-policy>` or :ref:`host the policy as a RESTful server <host-admin-policy-as-server>`.
2. Install the policy at the :ref:`server-side <server-side-admin-policy>` to enforce it for all users.
3. Optionally, an admin policy can also be installed at the :ref:`client-side<client-side-admin-policy>` if it needs to access user's local environment.

The order of policy application is demonstrated below:

> .. figure:: ../images/admin-policy.png

Reviewer: 1. there are a bit too many colors. can we only have color for the client and server policies?
2. Can we make the font size a bit larger?
3. Also, CLI should be CLI/SDK.

Author: Sounds good!

    :align: center
    :width: 80%
    :alt: Admin policy application order

.. note::

>     Client-side policy inherently lacks enforcement capability due to user ability to uninstall or modify them without administrative controls. But it is possible to use a client policy in conjunction with a server policy to enforce certain policies that can only be applied at client-side. Refer to :ref:`use-local-gcp-credentials-policy` for an example.

Reviewer: ```suggestion
    Client-side policy lacks enforcement capability, i.e., end-user may modify them. However, client policy is useful for automation that can only be applied at client-side. Refer to :ref:`use-local-gcp-credentials-policy` as an example.
```

.. _server-side-admin-policy:

Server-side
~~~~~~~~~~~

If you have a :ref:`centralized API server <sky-api-server>` deployed, you can enforce a policy for all users by setting it at the server-side.

.. tab-set::

>   .. tab-item:: Use a RESTful policy server

Reviewer: ```suggestion
  .. tab-item:: Use a RESTful policy
```

>     Open the server's dashboard, go to :ref:`the server's SkyPilot config <sky-api-server-config>` and set the :ref:`admin_policy <config-yaml-admin-policy>` field to the API URL of the policy server.

Reviewer: ```suggestion
    Open :ref:`SkyPilot dashboard <sky-api-server-config>` https://api.server.com/dashboard/config,  and set the :ref:`admin_policy <config-yaml-admin-policy>` field to the URL of the RESTful policy. To host a RESTful policy, see :ref:`here <host-admin-policy-as-server>`.
```

    .. code-block:: yaml

        admin_policy: mypackage.subpackage.MyPolicy

.. _client-side-admin-policy:

Client-side
~~~~~~~~~~~

If the policy needs to access user's local environment, you can get the policy applied at the client-side by following the steps below.

First, install the Python package that implements the policy:

.. code-block:: bash

    pip install mypackage.subpackage

Then, set the :ref:`admin_policy <config-yaml-admin-policy>` field in :ref:`the SkyPilot config <config-yaml>` to the path of the Python package that implements the policy.

.. code-block:: yaml

    admin_policy: mypackage.subpackage.MyPolicy

Optionally, you can also apply different policies in different projects by leveraging the :ref:`layered config <config-sources-and-overrides>`, e.g. set a different policy in ``$pwd/.sky.yaml`` for the current project:

.. code-block:: yaml

    admin_policy: mypackage.subpackage.AnotherPolicy

.. hint::

    SkyPilot loads the policy from the given package in the same Python environment.
    You can test the existence of the policy by running:

    .. code-block:: bash

        python -c "from mypackage.subpackage import MyPolicy"

.. note::

>     It is possible to call a RESTful policy at client-side. However, a RESTful policy is executed on the policy server host thus cannot access user's local environment inherently. If you want to apply a RESTful policy, it is recommended to install it at :ref:`server-side<server-side-admin-policy>` for better administrative control.

Reviewer: ```suggestion
    It is possible to call a RESTful policy at client-side. However, a RESTful policy is executed on the policy server host, i.e., cannot access user's local environment, e.g., local files.
```

    :pyobject: UseLocalGcpCredentialsPolicy
    :caption: `UseLocalGcpCredentialsPolicy <https://github.com/skypilot-org/skypilot/blob/master/examples/admin_policy/example_policy/example_policy/client_policy.py>`_

Specify the following config in :ref:`the SkyPilot config <config-yaml>` at the client:

.. literalinclude:: ../../../examples/admin_policy/use_local_gcp_credentials.yaml
    :language: yaml
    :caption: `Config YAML for using UseLocalGcpCredentialsPolicy <https://github.com/skypilot-org/skypilot/blob/master/examples/admin_policy/use_local_gcp_credentials.yaml>`_

Then specify the policy at the server with the same config, or call this policy in the :ref:`RESTful policy server <host-admin-policy-as-server>`.

> In the above example, the policy has to access the user's local environment to discover the user's GCP credentials and upload the credentials to the cluster by modifying the ``file_mounts`` field. So the key functionality of this policy must be executed as a client-side policy. In addition, when this policy is called at server-side, it will check whether a policy of the expected version has been applied at client-side and rejects the request if not. So the client-side application of the policy can be enforced on the server for non-malicious users.

Reviewer: ```suggestion
The policy above can access the user's local environment to discover the user's GCP credentials and upload the credentials to the cluster by modifying the ``file_mounts`` field. Note that the key functionality of this policy is executed as a client-side policy to discover the local credentials.

When this policy is called at server-side, it will check whether a policy of the expected version has been applied at client-side and rejects the request if not.
```

Author: This is much clearer, thanks!
